plugins/database/hana/hana.go

```diff
@@ -0,0 +1,284 @@
+	"github.com/hashicorp/vault/builtin/logical/database/dbplugin"
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/plugins"
+	"github.com/hashicorp/vault/plugins/helper/database/connutil"
+	"github.com/hashicorp/vault/plugins/helper/database/credsutil"
+	"github.com/hashicorp/vault/plugins/helper/database/dbutil"
+)
+
+const (
+	hanaTypeName = "hdb"
+)
+
+// HANA is an implementation of Database interface
+type HANA struct {
+	connutil.ConnectionProducer
+	credsutil.CredentialsProducer
+}
+
+// New implements builtinplugins.BuiltinFactory
+func New() (interface{}, error) {
+	connProducer := &connutil.SQLConnectionProducer{}
+	connProducer.Type = hanaTypeName
+
+	credsProducer := &credsutil.SQLCredentialsProducer{
+		DisplayNameLen: 32,
+		UsernameLen:    128,
+	}
+
+	dbType := &HANA{
+		ConnectionProducer:  connProducer,
+		CredentialsProducer: credsProducer,
+	}
+
+	return dbType, nil
+}
+
+// Run instantiates a HANA object, and runs the RPC server for the plugin
+func Run(apiTLSConfig *api.TLSConfig) error {
+	dbType, err := New()
+	if err != nil {
+		return err
+	}
+
+	plugins.Serve(dbType.(*HANA), apiTLSConfig)
+
+	return nil
+}
+
+// Type returns the TypeName for this backend
+func (h *HANA) Type() (string, error) {
+	return hanaTypeName, nil
+}
+
+func (h *HANA) getConnection() (*sql.DB, error) {
+	db, err := h.Connection()
+	if err != nil {
+		return nil, err
+	}
+
+	return db.(*sql.DB), nil
+}
+
+// CreateUser generates the username/password on the underlying HANA secret backend
+// as instructed by the CreationStatement provided.
+func (h *HANA) CreateUser(statements dbplugin.Statements, usernamePrefix string, expiration time.Time) (username string, password string, err error) {
+	// Grab the lock
+	h.Lock()
+	defer h.Unlock()
+
+	// Get the connection
+	db, err := h.getConnection()
+	if err != nil {
+		return "", "", err
+	}
+
+	if statements.CreationStatements == "" {
+		return "", "", dbutil.ErrEmptyCreationStatement
+	}
+
+	// Generate username
+	username, err = h.GenerateUsername(usernamePrefix)
+	if err != nil {
+		return "", "", err
+	}
+
+	// HANA does not allow hyphens in usernames, and highly prefers capital letters
+	username = strings.Replace(username, "-", "_", -1)
+	username = strings.ToUpper(username)
+
+	// Generate password
+	password, err = h.GeneratePassword()
+	if err != nil {
+		return "", "", err
+	}
+
+	// Most HANA configurations have password constraints.
+	// Prefix with A1a (user must change password upon login anyway)
+	password = strings.Replace(password, "-", "_", -1)
+	password = "A1a" + password
+
+	// expirationStr = (expiration - time.Now).Seconds() + HANA server side time
+	// This ensures hana will deactivate the user when time is up, regardless of vault's actions
+	var expirationStr string
+	timeQuery := fmt.Sprintf("SELECT TO_NVARCHAR(add_seconds(CURRENT_TIMESTAMP,"+
+		"%f), 'YYYY-MM-DD HH24:MI:SS') FROM DUMMY", (expiration.Sub(time.Now())).Seconds())
```

Comment:
Instead of adding the number of seconds you could just pass a formatted timestamp into the query, https://github.com/hashicorp/vault/blob/master/plugins/helper/database/credsutil/sql.go#L53 will format it correctly.